File: epi_judge_python_solutions/buy_and_sell_stock_k_times.py
from test_framework import generic_test


def buy_and_sell_stock_k_times(prices, k):

    if not k:
        return 0.0
    elif 2 * k >= len(prices):
        return sum(max(0, b - a) for a, b in zip(prices[:-1], prices[1:]))
    min_prices, max_profits = [float('inf')] * k, [0] * k
    for price in prices:
        for i in reversed(list(range(k))):
            max_profits[i] = max(max_profits[i], price - min_prices[i])
            min_prices[i] = min(min_prices[i],
                                price - (0 if i == 0 else max_profits[i - 1]))
    return max_profits[-1]


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main("buy_and_sell_stock_k_times.py",
                                       'buy_and_sell_stock_k_times.tsv',
                                       buy_and_sell_stock_k_times))

File: epi_judge_java_solutions/epi/BuyAndSellStockKTimes.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class BuyAndSellStockKTimes {
  @EpiTest(testDataFile = "buy_and_sell_stock_k_times.tsv")

  public static double buyAndSellStockKTimes(List<Double> prices, int k) {

    if (k == 0) {
      return 0.0;
    } else if (2 * k >= prices.size()) {
      return unlimitedPairsProfits(prices);
    }
    List<Double> minPrices =
                     new ArrayList<>(Collections.nCopies(k, Double.MAX_VALUE)),
                 maxProfits = new ArrayList<>(Collections.nCopies(k, 0.0));
    for (Double price : prices) {
      for (int i = k - 1; i >= 0; --i) {
        maxProfits.set(i,
                       Math.max(maxProfits.get(i), price - minPrices.get(i)));
        minPrices.set(i,
                      Math.min(minPrices.get(i),
                               price - (i > 0 ? maxProfits.get(i - 1) : 0.0)));
      }
    }
    return maxProfits.get(maxProfits.size() - 1);
  }

  private static double unlimitedPairsProfits(List<Double> prices) {
    double profit = 0.0;
    for (int i = 1; i < prices.size(); ++i) {
      profit += Math.max(0.0, prices.get(i) - prices.get(i - 1));
    }
    return profit;
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "BuyAndSellStockKTimes.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}
